File: tests/application/usecase/user/get_test.go
package user_test

import (
	"context"
	"testing"

	"github.com/stretchr/testify/assert"

	userusecase "go-clean-template/internal/application/usecase/user"
	"go-clean-template/internal/domain"
)

func TestGetByID_Success(t *testing.T) {
	mockRepo := new(MockUserRepo)
	uc := userusecase.NewGetUserUseCase(mockRepo)

	user := &domain.User{ID: "123", Name: "Alice", Email: "alice@example.com"}
	mockRepo.On("FindByID", "123").Return(user, nil)

	resp, err := uc.Execute(context.Background(), &userusecase.GetUserInput{ID: "123"})

	assert.NoError(t, err)
	assert.NotNil(t, resp)
	assert.NotNil(t, resp.User)
	assert.Equal(t, "123", resp.User.ID)
	assert.Equal(t, "Alice", resp.User.Name)

	mockRepo.AssertExpectations(t)
}

func TestGetByID_NotFound(t *testing.T) {
	mockRepo := new(MockUserRepo)
	uc := userusecase.NewGetUserUseCase(mockRepo)

	mockRepo.On("FindByID", "999").Return(nil, domain.ErrNotFound)

	resp, err := uc.Execute(context.Background(), &userusecase.GetUserInput{ID: "999"})

	assert.Error(t, err)
	assert.ErrorIs(t, err, domain.ErrNotFound)
	assert.Nil(t, resp)
	mockRepo.AssertExpectations(t)
}

File: tests/usecase/user/get_test.rs
use std::sync::Arc;

use async_trait::async_trait;

use rust_clean_template::application::usecase::user::{
    GetUserInput, GetUserUseCase, GetUserUseCaseImpl, UserRepository,
};
use rust_clean_template::domain;

struct MockUserRepo {
    result: Arc<std::sync::Mutex<Option<Result<domain::User, domain::Error>>>>,
}

#[async_trait]
impl UserRepository for MockUserRepo {
    async fn find_by_id(&self, _id: &str) -> Result<domain::User, domain::Error> {
        self.result
            .lock()
            .unwrap()
            .clone()
            .unwrap_or(Err(domain::Error::NotFound))
    }
    async fn find_all(
        &self,
        _offset: usize,
        _limit: usize,
    ) -> Result<(Vec<domain::User>, i64), domain::Error> {
        Ok((vec![], 0))
    }
    async fn find_by_email(&self, _email: &str) -> Result<domain::User, domain::Error> {
        Err(domain::Error::NotFound)
    }
    async fn create(&self, _user: &domain::User) -> Result<(), domain::Error> {
        Ok(())
    }
    async fn update(&self, _user: &domain::User) -> Result<(), domain::Error> {
        Ok(())
    }
    async fn delete(&self, _id: &str) -> Result<(), domain::Error> {
        Ok(())
    }
}

#[tokio::test]
async fn test_get_user_success() {
    let user = domain::User::new(
        "Alice".into(),
        "alice@example.com".into(),
        "123".into(),
        vec![],
    )
    .unwrap();
    let repo = MockUserRepo {
        result: Arc::new(std::sync::Mutex::new(Some(Ok(user)))),
    };

    let uc = GetUserUseCaseImpl::new(Arc::new(repo));
    let result = uc.execute(GetUserInput { id: "123".into() }).await;
    assert!(result.is_ok());
    assert_eq!(result.unwrap().user.name, "Alice");
}

#[tokio::test]
async fn test_get_user_not_found() {
    let repo = MockUserRepo {
        result: Arc::new(std::sync::Mutex::new(Some(Err(domain::Error::NotFound)))),
    };

    let uc = GetUserUseCaseImpl::new(Arc::new(repo));
    let result = uc.execute(GetUserInput { id: "999".into() }).await;
    assert!(result.is_err());
}
